### ISSUE TITLE
Spring Cloud Sleuth GCP: every time spans are sent, they are sent to localhost as well

### ISSUE THREAD

**Issue by bbzg** [spring-cloud/spring-cloud-gcp#1705]
There should be a way to turn off the default ZipkinRestTemplateSenderConfiguration when using Spring Cloud GCP. I would prefer if it was turned off by default.

Currently, every time spans are being sent, a completely unnecessary call to `http://localhost:9411/api/v2/spans`, and it fails 100% of the time because there is nothing there to answer.

The only reason I saw it was because I used `logging.level.org.springframework.cloud=TRACE`.

I tried to fix it by setting `spring.zipkin.enabled=false` but it crashes on startup because of `ZipkinBackwardsCompatibilityAutoConfiguration`.

If there already is a way to turn this off I would be very happy. Anyway my opinion is it should be off by default because I have already chosen `spring.cloud.gcp.trace.enabled=true`.

Originally posted as https://github.com/spring-cloud/spring-cloud-sleuth/issues/1377 but probably better to originate a fix in this repo instead.

Thanks.

**Comment by meltsufin** [comment]
What is the purpose of `ZipkinRestTemplateSenderConfiguration` and how can one turn it off?@marcingrzejszczak

**Comment by saturnism** [comment]
/cc @TYsewyn 

Was discussing w/ @marcingrzejszczak in person. We think this is what's happening:
1. [TraceAutoConfiguration](https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java) is picking up multiple `Reporter`s.
1. [ZipkinAutoConfiguration](https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-zipkin/src/main/java/org/springframework/cloud/sleuth/zipkin2/ZipkinAutoConfiguration.java) configures a Zipkin Reporter that's conditional on the Zipkin Reporter's bean name.
1. [StackdriverTraceAutoConfiguration](https://github.com/spring-cloud/spring-cloud-gcp/blame/master/spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/trace/StackdriverTraceAutoConfiguration.java) configures Stackdriver's Zipkin Reporter, using its own bean name.

As a result, both reporters are picked up. During Sleuth's 2.1 RC releases, Stackdriver autoconfiguration used the same bean name as Zipkin, so that there is no duplication. For Sleuth's 2.1 GA release, the bean name changed and so now we have duplicate.

For @marcingrzejszczak and @TYsewyn: Stackdriver Zipkin Reporter is meant to use existing Zipkin/Brave integrations (e.g., RestTemplate, etc), but only replacing the reporter.  Would it make sense to simply override the existing Zipkin sender/reporter bean names?

**Comment by saturnism** [comment]
@bbzg could i trouble to share the pom? It would help to know what BOMs you are using and what direct dependencies you have as well (and the versions)

**Comment by TYsewyn** [comment]
I was about to ask the same, because I think this might already be solved with Greenwich.SR1. 🤔

**Comment by marcingrzejszczak** [comment]
Related issue - https://github.com/spring-cloud/spring-cloud-sleuth/pull/1192

**Comment by saturnism** [comment]
Follow up question is - what did it have both Zipkin Sender and Stackdriver Sender so that one of them needed to be disabled? According to @marcingrzejszczak that shouldn't be happening. We are suspecting somewhere the `spring-cloud-sleuth-zipkin` is being pulled in. Seeing the pom.xml would help.

**Comment by TYsewyn** [comment]
@saturnism this is indeed the case when you're using GCP 1.1.x or older - we couldn't align the changes in Spring Cloud Sleuth core with the ones made here for the Greenwich release.
It's the auto configuration module of GCP that's pulling in this dependency.

**Comment by saturnism** [comment]
Yeah, just confirmed that `spring-cloud-gcp-starter-trace` is pulling in `spring-cloud-sleuth-zipkin`, and thus the duplicate sender.  Up until all the `spring-cloud-sleuth` 2.1.0.RCx we've been overriding the sender (i.e., only 1 sender). But it seems like the multi-sender was changed just for the actual 2.1.0 RELEASE. With multi-sender, we should remove `spring-cloud-sleuth-zipkin` dependency from our starter. It looks like it's already done in `master`.

In the mean time, I tried to exclude `spring-cloud-sleuth-zipkin` from the dependencies:
```
<dependency>
  <groupId>org.springframework.cloud</groupId>
  <artifactId>spring-cloud-gcp-starter-trace</artifactId>
  <exclusions>
    <exclusion>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-sleuth-zipkin</artifactId>
    </exclusion>
  </exclusions>
</dependency>
```

However, this not only disabled Zipkin sender to localhost, but also disabled sender to Stackdriver Trace, because our autoconfigure only configures the `Sender` and not the `Reporter`. This worked for all the Spring Cloud Sleuth 2.1 RCs.

I'm assuming the backwards compatibility stuff should've:
1. take the sender `spring-cloud-gcp` produced
1. produce the zipkin reporter and so there is no duplicate reporters

This doesn't seem to be working. `ZipkinBackwardsCompatibilityAutoConfiguration` is producing the reporter [code](https://github.com/spring-cloud/spring-cloud-sleuth/blob/2.1.x/spring-cloud-sleuth-zipkin/src/main/java/org/springframework/cloud/sleuth/zipkin2/ZipkinBackwardsCompatibilityAutoConfiguration.java#L80), however it's doesn't seem to be using the correct bean name.

To validate this assumption, @marcingrzejszczak and I tried a workaround:
1. Exclude `spring-cloud-sleuth-zipkin` as before.
1. Produce our own Reporter
   ```
   @Bean
   public Reporter<Span> reporter(ReporterMetrics reporterMetrics,
      @Qualifier("stackdriverSender") Sender sender) {
        return AsyncReporter.builder(sender).queuedMaxSpans(1000)
          .messageTimeout(1, TimeUnit.SECONDS)
          .metrics(reporterMetrics).build(StackdriverEncoder.V1);
    }
    ```

**Comment by TYsewyn** [comment]
I created a PR for the 1.1.x branch to verify the expected behaviour with the `ZipkinBackwardsCompatibilityAutoConfiguration` in place.
This should confirm that @bbzg is using an older version of Spring Cloud.

**Comment by meltsufin** [comment]
It looks like we just need to backport #1383 to the `1.1.x` branch.

**Comment by TYsewyn** [comment]
That would indeed be the conclusion of this https://github.com/spring-cloud/spring-cloud-sleuth/issues/1400#issuecomment-512231850

**Comment by meltsufin** [comment]
We'd be happy to accept another contribution form you, since you submitted the original PR!

**Comment by bbzg** [comment]
@saturnism This is what I use:

```
allprojects {
    apply plugin: "io.spring.dependency-management"

    dependencyManagement {
        imports {
            mavenBom 'org.springframework.boot:spring-boot-dependencies:2.1.5.RELEASE'
            mavenBom 'com.google.cloud:google-cloud-bom:0.94.0-alpha'
            mavenBom 'org.springframework.cloud:spring-cloud-dependencies:Greenwich.RELEASE'
        }

    // ...
```

**Comment by TYsewyn** [comment]
Oh yeah, you might want to upgrade Spring Boot to 2.1.6.RELEASE and Spring Cloud to Greenwich.SR2. :)

In the meantime we backported this functionality from 1.2.x to 1.1.x.
If you upgrade Spring Cloud you should be able to disable Zipkin with `spring.zipkin.enabled=false`.

**Comment by saturnism** [comment]
awesome! thanks all for the help! I suppose on GCP side, we need to upgrade the dependency version for the next 1.1.x revision release.

**Comment by TYsewyn** [comment]
No, it should be good for this issue. We changed things on GCP side, not on Sleuth side.

**Comment by meltsufin** [comment]
Fixed by #1744
